cmd/zc_traverser_blob_versions.go

```diff
@@ -0,0 +1,130 @@
+	includeDirectoryStubs       bool
+	incrementEnumerationCounter enumerationCounterFunc
+	listOfVersionIds            []string
+}
+
+func (t *blobVersionsTraverser) isDirectory(isSource bool) bool {
+	isDirDirect := copyHandlerUtil{}.urlIsContainerOrVirtualDirectory(t.rawURL)
+
+	// Skip the single blob check if we're checking a destination.
+	// This is an individual exception for blob because blob supports virtual directories and blobs sharing the same name.
+	if isDirDirect || !isSource {
+		return isDirDirect
+	}
+
+	_, err := t.getBlobProperties("")
+
+	return err != nil
+}
+
+func (t *blobVersionsTraverser) getBlobProperties(versionID string) (props *azblob.BlobGetPropertiesResponse, err error) {
+	blobURLParts := azblob.NewBlobURLParts(*t.rawURL)
+	blobURLParts.BlobName = strings.TrimSuffix(blobURLParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)
+	if versionID != "" {
+		blobURLParts.VersionID = versionID
+	}
+
+	blobURL := azblob.NewBlobURL(blobURLParts.URL(), t.p)
+	props, err = blobURL.GetProperties(t.ctx, azblob.BlobAccessConditions{})
+
+	if err != nil {
+		return props, err
+	}
+
+	// if there was no problem getting the properties, it implies that we are looking at a single blob
+	if gCopyUtil.doesBlobRepresentAFolder(props.NewMetadata()) {
+		return props, errors.New("This is not a blob")
+	}
+	return props, nil
+
+}
+
+func (t *blobVersionsTraverser) traverse(preprocessor objectMorpher, processor objectProcessor, filters []objectFilter) (err error) {
+	blobURLParts := azblob.NewBlobURLParts(*t.rawURL)
+
+	for _, versionID := range t.listOfVersionIds {
+		blobProperties, err := t.getBlobProperties(versionID)
+
+		if err != nil {
+			return err
+		}
+
+		if blobProperties == nil {
+			panic("isBlob should never be set if getting properties is an error")
+		}
+		blobURLParts.VersionID = versionID
+		storedObject := newStoredObject(
+			preprocessor,
+			getObjectNameOnly(strings.TrimSuffix(blobURLParts.BlobName, common.AZCOPY_PATH_SEPARATOR_STRING)),
+			"",
+			common.EEntityType.File(),
+			blobProperties.LastModified(),
+			blobProperties.ContentLength(),
+			blobProperties,
+			blobPropertiesResponseAdapter{blobProperties},
+			common.FromAzBlobMetadataToCommonMetadata(blobProperties.NewMetadata()),
+			blobURLParts.ContainerName,
+			versionID,
+		)
+
+		if t.incrementEnumerationCounter != nil {
+			t.incrementEnumerationCounter(common.EEntityType.File())
+		}
+
+		err = processIfPassedFilters(filters, storedObject, processor)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func newBlobVersionsTraverser(rawURL *url.URL, p pipeline.Pipeline, ctx context.Context, recursive, includeDirectoryStubs bool,
+	incrementEnumerationCounter enumerationCounterFunc, listOfVersionIds []string) (t *blobVersionsTraverser) {
+	return &blobVersionsTraverser{
+		rawURL:                      rawURL,
+		p:                           p,
+		ctx:                         ctx,
+		recursive:                   recursive,
```

Comment:
>recursive [](start = 31, length = 9)

recursive doesn't apply to this traverser, right?

Reply:
Ohh correct. Since the versions are already specified